pkg/k8s/cep.go

```diff
@@ -0,0 +1,363 @@
+		return
+	}
+	if !ciliumEPControllerLimit.Check(sv) {
+		scopedLog.WithFields(logrus.Fields{
+			"expected": sv,
+			"found":    ciliumEPControllerLimit,
+		}).Warn("cannot run with this k8s version")
+		return
+	}
+
+	ciliumClient, err := getCiliumClient()
+	if err != nil {
+		scopedLog.WithError(err).Error("Not starting controller because unable to get cilium k8s client")
+		return
+	}
+	k8sClient := Client()
+
+	// this dummy manager is needed only to add this controller to the global list
+	controller.NewManager().UpdateController(controllerName,
+		controller.ControllerParams{
+			RunInterval: 30 * time.Minute,
+			DoFunc: func() error {
+				// Don't run immediately
+				if firstRun {
+					firstRun = false
+					return nil
+				}
+
+				// Only run if we are the "lowest" node in our cluster, when sorted by
+				// names. This means only one node will ever run GCs in a given
+				// cluster.
+				thisNode := node.GetLocalNode().Identity()
+				thisNodeStr := thisNode.String()
+				for id := range node.GetNodes() {
+					if idStr := id.String(); id.Cluster == thisNode.Cluster && idStr < thisNodeStr {
+						scopedLog.WithField(logfields.Node, idStr).Debug("Skipping GC because another node is a better candidate")
+						return nil
+					}
+				}
+
+				clusterPodSet := map[string]bool{}
+				clusterPods, err := k8sClient.CoreV1().Pods("").List(meta_v1.ListOptions{})
+				if err != nil {
+					return err
+				}
+				for _, pod := range clusterPods.Items {
+					podFullName := pod.Name + ":" + pod.Namespace
+					clusterPodSet[podFullName] = true
+				}
+
+				// "" is all-namespaces
+				ceps, err := ciliumClient.CiliumEndpoints(meta_v1.NamespaceAll).List(meta_v1.ListOptions{})
+				if err != nil {
+					scopedLog.WithError(err).Debug("Cannot list CEPs")
+					return err
+				}
+				for _, cep := range ceps.Items {
+					cepFullName := cep.Name + ":" + cep.Namespace
+					if _, found := clusterPodSet[cepFullName]; !found {
+						// delete
+						scopedLog = scopedLog.WithFields(logrus.Fields{
+							logfields.EndpointID: cep.Status.ID,
+							logfields.K8sPodName: cepFullName,
+						})
+						scopedLog.Debug("Orphaned CiliumEndpoint is being garbage collected")
+						if err := ciliumClient.CiliumEndpoints(cep.Namespace).Delete(cep.Name, &meta_v1.DeleteOptions{}); err != nil {
+							scopedLog.WithError(err).Debug("Unable to delete CEP")
+							return err
+						}
+					}
+				}
+				return nil
+			},
+		})
+}
+
+// RunK8sCiliumEndpointSync starts a controller that synchronizes the endpoint
+// to the corresponding k8s CiliumEndpoint CRD
+// CiliumEndpoint objects have the same name as the pod they represent
+func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoint) {
+	var (
+		endpointID     = e.ID
+		controllerName = fmt.Sprintf("sync-to-k8s-ciliumendpoint (%v)", endpointID)
+		scopedLog      = e.GetLogger().WithField("controller", controllerName)
```

Comment:
Please use `e.Logger(subsystem)` instead. This avoids the need for exporting `getLogger()`, and provides the context of where the log message actually came from rather than misleading that it comes from the endpoint package.

Reply:
Ah, I didn't know that function existed. Dropped the `getLogger` --> `GetLogger` commit in next upload and changed this function to use `Logger`.